### PR TITLE
Add preprocess for Dictionaries

### LINKED ISSUE #47: Unpacking/Preprocess-ing of Dictionaries
Following #46 I realised that Optimums stores optimisation metadata in Dictionaries.

Right now, if a dictionary is logged to `TBLogger` (or if it's found while unpacking other structs) it is treated like any other struct, and it's internal fields are recursively logged:
```
Julia> using TensorBoardLogger
julia> data = TensorBoardLogger.preprocess("dict", Dict(:a=>1, :b=>2), [])
8-element Array{Any,1}:
    "dict/slots" => UInt8[0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
     "dict/keys" => Symbol[:a, :b, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef, #undef]
     "dict/vals" => [1, 2, 32, 1, 11, 2, 4628399872, 108, 24, 4628399872, 4628400112, 4628402464, 4628400112, 4628398656, 4628400112, 4426404768]
     "dict/ndel" => 0
    "dict/count" => 2
      "dict/age" => 0x0000000000000002
 "dict/idxfloor" => 1
 "dict/maxprobe" => 0
```

I think we should treat dictionaries with `String` and `Symbol` keys as we treat structures, and further preprocess key-value pairs.

Although, people often don't specialise the key type and use `Dict{Any,Any}` (Optim does that).
In this case, we could check if a string representation of every key exists, otherwise log them both as text?

### PR DESCRIPTION
**Pull request by PhilipVinc**
Fixes #47 

Should allow us to log almost by default what `Optim` throws at us (and a few other things)

### CODE REVIEW

File: src/logger_dispatch.jl [hunk 1]
```diff
@@ -38,6 +38,22 @@ See also: [`Base.propertynames`](@ref)
 """
 logable_propertynames(val::Any) = propertynames(val)
 
+## Default unpacking of key-value dictionaries
+function preprocess(name, dict::Dict, data)
```
Comment (oxinabox) [inline src/logger_dispatch.jl]:
```suggestion
function preprocess(name, dict::AbstractDict, data)
```

File: src/logger_dispatch.jl [hunk 2]
```diff
@@ -38,6 +38,22 @@ See also: [`Base.propertynames`](@ref)
...
+    i=1
+    for k=keys(dict)
```
Comment (oxinabox) [inline src/logger_dispatch.jl]: Can use `enumerate` for this.
Then can delete the `i=1` and `i += 1`
```suggestion
    for (i,k) in enumerate(keys(dict))
```

File: src/logger_dispatch.jl [hunk 3]
```diff
@@ -38,6 +38,22 @@ See also: [`Base.propertynames`](@ref)
...
+        if isa(k, String) || isa(k, Symbol) || isa(k, Real)
```
Comment (oxinabox) [inline src/logger_dispatch.jl]:
```suggestion
        if isa(k, AbstractString) || isa(k, Symbol) || isa(k, Real)
```

**Comment by oxinabox** [comment]
How about this alternative:

```
function preprocess(name, dict::AbstractDict, data)
    for (key, val) in dict
        # convert any key into a string, via interpolating it
        preprocess("$name/$key", val, data)
    end
    return data
end
```

Interpolation, calls `string` under the hood,
and `string` will convert *anything* into a string (it in turn calls `print` under the hood)
I think if you used something as a key in a dict and that dict came to a logger, you wanted to having it stringified via the mechanisms provided.

```
julia> struct Foo x end

julia> f= Foo(10)
Foo(10)

julia> string(f);

julia> string(Foo(Foo(f)))
"Foo(Foo(Foo(10)))"
```

**Comment by PhilipVinc** [comment]
That's what I had fought at the beginning, but that I started thinking about weird dictionaries (like `IdDict` from Zygote) and what would happen if we let users serialise those.

But indeed, the idea of a dictionary is to have keys to access values, so... I guess what you are proposing is a better way to handle everything

OTOH, this remembers me that we should start doing `tag` sanitisation. TensorBoard has some limits in what it can handle in the tag. I'm pretty sure it cannot handle unicode, so we should keep that in mind.

**Comment by oxinabox** [comment]
Even with an `IdDict` I would expect this behavour,
of only logging the values, indexed by the `string`ified keys.

**Review by oxinabox — APPROVED** [review]

**Comment by xukai92** [comment]
I think similar thing should be added to `NamedTuple`?

**Comment by oxinabox** [comment]
NamedTuple is covered by our default destructuring for now at least

**Comment by xukai92** [comment]
Oh I didn't realize this. Thanks!

**Pull request merged**